### PR TITLE
[ELY-338] Realm selection

### PR DESCRIPTION
**Pull request by dmlloyd**

### CODE REVIEW

**Comment by dmlloyd** [comment]
This should get us to the point where we can create SASL servers and HTTP mechanisms with realm selection.  The realm and mechanism configuration is not yet reflected in the SASL/HTTP configurations, but if this PR looks OK I have some ideas for the best way to represent that.

File: src/main/java/org/wildfly/security/auth/server/AbstractMechanismAuthenticationFactory.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                }
+            }
+            if (supported) {
+                names.add(mechName);
+            }
+        }
+        return names;
+    }
+
+    abstract Collection<String> getAllSupportedMechNames();
```
Comment (darranl) [inline src/main/java/org/wildfly/security/auth/server/AbstractMechanismAuthenticationFactory.java]: Maybe 'MechanismNames' ?  Previous method is in that form.

Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/AbstractMechanismAuthenticationFactory.java]: It's not a public class so I wasn't too concerned about it.

File: src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java [hunk 2]
```diff
@@ -533,9 +434,10 @@ public boolean authorize(String name) throws IllegalArgumentException, RealmUnav
 
     /**
      * Mark this authentication as "successful".  The context cannot be used after this method is called, however
-     * the authorized identity may thereafter be accessed via the {@link #getAuthorizedIdentity()} method.
+     * the authorized identity may thereafter be accessed via the {@link #getAuthorizedIdentity()} method.  If no
+     * authentication actually happened, then authentication will complete anonymously.
      *
-     * @throws IllegalStateException if no authentication has been initiated or authentication is already completed
+     * @throws IllegalStateException if authentication is already completed
```
Comment (pskopek) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: Minor: There is @throws RealmUnavailableException from javadoc

**Pull request merged**